service/common/src/main/java/org/apache/polaris/service/events/listeners/InMemoryBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,166 @@
+  private final Duration timeToFlush;
+  private final int maxBufferSize;
+
+  @Context ContainerRequestContext containerRequestContext;
+
+  private record EventAndContext(PolarisEvent polarisEvent, PolarisCallContext callContext) {}
+
+  @Inject
+  public InMemoryBufferPolarisPersistenceEventListener(
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      Clock clock,
+      InMemoryBufferPersistenceListenerConfiguration eventListenerConfiguration) {
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.clock = clock;
+    this.timeToFlush =
+        eventListenerConfiguration.bufferTime().orElse(Duration.of(30, ChronoUnit.SECONDS));
+    this.maxBufferSize = eventListenerConfiguration.maxBufferSize().orElse(5); // 5 events default
+  }
+
+  @PostConstruct
+  void start() {
+    futures.put(
+        executor.scheduleAtFixedRate(
+            this::runCleanup, 0, timeToFlush.toMillis(), TimeUnit.MILLISECONDS),
+        1);
+  }
+
+  void runCleanup() {
+    for (String realmId : buffer.keySet()) {
+      try {
+        checkAndFlushBufferIfNecessary(realmId);
+      } catch (Exception e) {
+        LOGGER.debug("Buffer checking task failed for realm ({}): {}", realmId, e);
+      }
+    }
+    // Clean up futures
+    try {
+      futures.keySet().removeIf(future -> future.isCancelled() || future.isDone());
+    } catch (Exception e) {
+      LOGGER.debug("Futures reaper task failed.");
+    }
+  }
+
+  @PreDestroy
+  void shutdown() {
+    futures.keySet().forEach(future -> future.cancel(false));
+    executor.shutdownNow();
+  }
+
+  @Override
+  String getRequestId() {
+    if (containerRequestContext != null && containerRequestContext.hasProperty(REQUEST_ID_KEY)) {
+      return (String) containerRequestContext.getProperty(REQUEST_ID_KEY);
+    }
+    return UUID.randomUUID().toString();
+  }
+
+  @Override
+  void addToBuffer(PolarisEvent polarisEvent, CallContext callCtx) {
+    String realmId = callCtx.getRealmContext().getRealmIdentifier();
+
+    buffer
+        .computeIfAbsent(realmId, k -> new ConcurrentLinkedQueue<>())
+        .add(new EventAndContext(polarisEvent, callCtx.getPolarisCallContext().copy()));
+    futures.put(executor.submit(() -> checkAndFlushBufferIfNecessary(realmId)), 1);
+  }
+
+  @VisibleForTesting
+  public void checkAndFlushBufferIfNecessary(String realmId) {
+    ConcurrentLinkedQueue<EventAndContext> queue = buffer.get(realmId);
+    if (queue == null || queue.isEmpty()) {
+      return;
+    }
+
+    // Given that we are using a ConcurrentLinkedQueue, this should not lock any calls to `add` on
+    // the queue.
+    synchronized (queue) {
+      // Double-check inside synchronized block
+      if (queue.isEmpty()) {
+        return;
+      }
+
+      EventAndContext head = queue.peek();
+      if (head == null) {
+        return;
+      }
+
+      Duration elapsed = Duration.ofMillis(clock.millis() - head.polarisEvent.getTimestampMs());
```

Comment:
`clock.millis() - head.polarisEvent.getTimestampMs()` relies on the wall-clock, not the actual elapsed time.

Reply:
We use `Clock::millis` all over the place including in persistence retry logic, and I don't see any reason that it's especially problematic here